sdks/python/apache_beam/runners/interactive/background_caching_job.py

```diff
@@ -75,14 +77,16 @@ def is_background_caching_job_needed(user_pipeline):
   return (has_source_to_cache(user_pipeline) and
           # Checks if it's the first time running a job from the pipeline.
           (not background_caching_job_result or
-           # Or checks if there is no valid previous job.
+           # Or checks if there is no previous job.
            background_caching_job_result.state not in (
                # DONE means a previous job has completed successfully and the
                # cached events are still valid.
                runners.runner.PipelineState.DONE,
                # RUNNING means a previous job has been started and is still
                # running.
-               runners.runner.PipelineState.RUNNING)))
+               runners.runner.PipelineState.RUNNING) or
+           # Or checks if we can invalidate the previous job.
+           is_unbounded_source_changed(user_pipeline)))
```

Comment:
Similar to the reason why we use has_source_to_cache() above instead of has_unbounded_source(), we need to see whether any of the sources to cache has changed. So perhaps we should change this to something like cache_sources_changed()

Reply:
Yes, I agree. Changing it into `is_source_to_cache_changed`.

I was thinking about change in a bounded source wouldn't affect cached unbounded sources. But it feels like that is going to split background caching job into 2 categories or make the instrumenting process complicated (when we add support to cache arbitrary source). Let's unify the source caching.